packages/core/content-manager/server/src/history/services/lifecycles.ts

```diff
@@ -27,110 +107,58 @@ const createLifecyclesService = ({ strapi }: { strapi: Core.Strapi }) => {
-         * When a document is published, the draft version of the document is also updated.
-         * It creates confusion for users because they see two history versions each publish action.
-         * To avoid this, we silence the update action during a publish request,
-         * so that they only see the published version of the document in the history.
-         */
-        if (
-          context.action === 'update' &&
-          strapi.requestContext.get()?.request.url.endsWith('/actions/publish')
-        ) {
-          return next();
-        }
+      strapi.documents.use(async (context, next) => {
+        const result = (await next()) as any;
 
-        const contentTypeUid = context.contentType.uid;
-        // Ignore content types not created by the user
-        if (!contentTypeUid.startsWith('api::')) {
-          return next();
+        if (!shouldCreateHistoryVersion(context)) {
+          return result;
         }
 
-        const result = (await next()) as any;
-
-        const documentContext = {
-          documentId:
-            context.action === 'create' || context.action === 'clone'
-              ? result.documentId
-              : context.params.documentId,
-          locale: context.params?.locale,
-        };
+        // On create/clone actions, the documentId is not available before creating the action is executed
+        const documentId =
+          context.action === 'create' || context.action === 'clone'
+            ? result.documentId
+            : context.params.documentId;
 
+        // Apply default locale if not available in the request
         const defaultLocale = await serviceUtils.getDefaultLocale();
-        const locale = documentContext.locale || defaultLocale;
-
-        if (Array.isArray(locale)) {
-          strapi.log.warn(
-            '[Content manager history middleware]: An array of locales was provided, but only a single locale is supported for the findOne operation.'
-          );
-          // TODO calls picked from the middleware could contain an array of
-          // locales. This is incompatible with our call to findOne below.
-          return next();
+        const locales = castArray(context.params?.locale || defaultLocale);
+        if (!locales.length) {
+          return result;
         }
 
-        const document = await strapi.documents(contentTypeUid).findOne({
-          documentId: documentContext.documentId,
-          locale,
-          populate: serviceUtils.getDeepPopulate(contentTypeUid),
+        // All schemas related to the content type
+        const uid = context.contentType.uid;
+        const schemas = getSchemas(uid);
+
+        // Find all affected entries
+        const localeEntries = await strapi.db.query(uid).findMany({
+          where: {
+            documentId,
+            locale: { $in: locales },
+            publishedAt: null,
+          },
+          populate: serviceUtils.getDeepPopulate(uid, true /* use database syntax */),
         });
```

Comment:
Why do you use the query service and not the document one?

Reply:
atm we don't have a way to get multiple locales of the same document, so this way we can save some queries we send to the db